repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java

```diff
@@ -0,0 +1,211 @@
+            SELECT id, name, description, price, duration, create_date, last_update_date
+            FROM gift_certificates
+            WHERE id = :id;
+            """;
+
+    private static final String INSERT = """
+            INSERT INTO gift_certificates (name, description, price, duration, create_date, last_update_date)
+            VALUES (:name, :description, :price, :duration, :create_date, :last_update_date);
+            """;
+
+    private static final String UPDATE = """
+            UPDATE gift_certificates
+            SET name = :name, description = :description, price = :price, duration = :duration, last_update_date = :last_update_date
+            WHERE id = :id;
+            """;
+
+    private static final String DELETE = """
+            DELETE FROM gift_certificates
+            WHERE id = :id;
+            """;
+
+    private static final String DELETE_FOREIGN_KEY = """
+            DELETE FROM certificates_tags
+            WHERE certificate_id = :id;
+            """;
+
+    private static final String INSERT_CERTIFICATES_TAGS_RELATION = """
+            INSERT INTO certificates_tags (certificate_id, tag_id)
+            VALUES (:certificate_id, :tag_id);
+            """;
+
+    private static final String DELETE_CERTIFICATES_TAGS_RELATION = """
+            DELETE FROM certificates_tags
+            WHERE certificate_id = :certificate_id AND tag_id = :tag_id
+            """;
+
+    private final NamedParameterJdbcTemplate jdbcTemplate;
+    private final RowMapper<GiftCertificate> rowMapper;
+
+    public GiftCertificateRepositoryImpl(DataSource dataSource, RowMapper<GiftCertificate> rowMapper) {
+        this.jdbcTemplate = new NamedParameterJdbcTemplate(dataSource);
+        this.rowMapper = rowMapper;
+    }
+
+    @Override
+    public List<GiftCertificate> filter(String tagName, String name, String description, String sort) {
+        MapSqlParameterSource parameters = new MapSqlParameterSource();
+        String queryString = buildFilterQuery(tagName, name, description, sort, parameters);
+        return jdbcTemplate.query(queryString, parameters, rowMapper);
+    }
+
+    @Override
+    public Optional<GiftCertificate> findById(long id) {
+        SqlParameterSource parameters = new MapSqlParameterSource().addValue(ID_PARAM, id);
+        List<GiftCertificate> giftCertificates = jdbcTemplate.query(SELECT_BY_ID, parameters, rowMapper);
+        return Optional.ofNullable(giftCertificates.size() == 1 ? giftCertificates.get(0) : null);
+    }
+
+    @Override
+    public void addRelation(long certificateId, long tagId) {
+        SqlParameterSource parameters = new MapSqlParameterSource()
+                .addValue(CERTIFICATE_ID_PARAM, certificateId)
+                .addValue(TAG_ID_PARAM, tagId);
+        jdbcTemplate.update(INSERT_CERTIFICATES_TAGS_RELATION, parameters);
+    }
+
+    @Override
+    public void removeRelation(long certificateId, long tagId) {
```

Comment:
make the name of the method more specific. removeRelation(long, long) is super abstract. From the first glance it's difficult to realize what relation is this method about. 
+ what would you do if certificate has 2/3 relations?

Reply:
fixed this issue